pilot/pkg/networking/v1alpha3/clusters.go

```diff
@@ -0,0 +1,210 @@
+func BuildClusters(env model.Environment, proxy model.Proxy) []*v2.Cluster {
+	clusters := make([]*v2.Cluster, 0)
+
+	// TODO node type router
+
+	// node type Sidecar
+	services, err := env.Services()
+	if err != nil {
+		log.Errorf("Failed for retrieve services: %v", err)
+		return nil
+	}
+
+	// build outbound clusters
+	for _, service := range services {
+		config := env.IstioConfigStore.DestinationRule(service.Hostname, "")
+		if config != nil {
+			destinationRule := config.Spec.(*networking.DestinationRule)
+			for _, svcPort := range service.Ports {
+				var hosts []*core.Address
+				if service.Resolution == model.DNSLB {
+					// FIXME port name not required if only one port
+					instances, err := env.ServiceDiscovery.Instances(service.Hostname, []string{svcPort.Name}, nil)
+					if err != nil {
+						log.Errorf("failed to retrieve instances for %s: %v", service.Hostname, err)
+					} else {
+						for _, instance := range instances {
+							hosts = append(hosts, &core.Address{Address: &core.SocketAddress{
+								Address:  instance.Endpoint.Address,
+								Protocol: core.TCP,
+								PortSpecifier: &core.SocketAddress_PortValue{
+									PortValue: uint32(instance.Endpoint.Port),
+								},
+							}})
+						}
+					}
+				}
+
+				// create default cluster
+				cluster := &v2.Cluster{
+					Name:  buildClusterName(service.Hostname, "", svcPort.Name),
+					Type:  convertResolution(service.Resolution),
+					Hosts: hosts,
+				}
+				applyTrafficPolicy(cluster, destinationRule.TrafficPolicy)
+				clusters = append(clusters, cluster)
+
+				for _, subset := range destinationRule.Subsets {
+					cluster := &v2.Cluster{
+						Name:  buildClusterName(service.Hostname, subset.Name, svcPort.Name),
+						Type:  convertResolution(service.Resolution),
+						Hosts: hosts,
+					}
+					applyTrafficPolicy(cluster, destinationRule.TrafficPolicy)
+					applyTrafficPolicy(cluster, subset.TrafficPolicy)
+
+					clusters = append(clusters, cluster)
+				}
+
+				// TODO HTTP2 feature
+
+			}
+		}
+	}
+
+	// TODO build inbound clusters
+	instances, err := env.GetProxyServiceInstances(proxy)
+	for _, instance := range instances {
+		clusters = append(clusters, &v2.Cluster{
+			Name: "",
+			Type: v2.Cluster_STATIC,
+			ConnectTimeout: 0,
+			LbPolicy: v2.Cluster_ROUND_ROBIN,
+			Hosts: []*core.Address{{Address: &core.SocketAddress{
+				Address: instance.Endpoint.Address,
+				Protocol: core.TCP,
+				PortSpecifier: &core.SocketAddress_PortValue{
+					PortValue: uint32(instance.Endpoint.Port),
+				},
+			}}},
+		})
+		// TODO: HTTP2 feature
+	}
+
+	// TODO add original dst cluster
+
+	return clusters // TODO: guaranteed ordering?
+}
+
+// TODO port name is not mandatory if only one port defined
+// TODO subset name can be empty
+func buildClusterName(hostname, subsetName, portName string) string {
+	return fmt.Sprintf("%s.%s|%s", subsetName, hostname, portName)
```

Comment:
I suggest `[in|out] | subsetName | fqdn | portname` for the cluster names.. if subset name is missing or if we are referring to the entire service, that field would be empty, i.e. "out | | hello.default.svc.cluster.local | http" , so that we can easily split the service key string by "|", and obtain the server + subset + port.